src/PaymentMethods/BuyNowPayLater/AbstractBuyNowPayLater.php

```diff
@@ -0,0 +1,509 @@
+			);
+		} catch ( \Exception $e ) {
+			wc_get_logger()->error( $e->getMessage() );
+			throw new \Exception( Utils::map_response_code_to_friendly_message() );
+		}
+	}
+
+	public function process_refund( $order_id, $amount = null, $reason = '' ) {
+		return $this->gateway->process_refund( $order_id, $amount, $reason );
+	}
+
+	/**
+	 * Initiate the payment.
+	 *
+	 * @param WC_Order $order
+	 *
+	 * @throws \GlobalPayments\Api\Entities\Exceptions\ApiException
+	 */
+	private function initiate_payment( WC_Order $order ) {
+		$request = $this->gateway->prepare_request( AbstractGateway::TXN_TYPE_BNPL_AUTHORIZE, $order );
+		$request->set_request_data( array(
+			'globalpayments_bnpl' => $this->get_provider_endpoints(),
+		) );
+
+		$gateway_response = $this->gateway->client->submit_request( $request );
+
+		$this->gateway->handle_response( $request, $gateway_response );
+
+		return $gateway_response;
+	}
+
+	/**
+	 * Handle customer redirect URL.
+	 */
+	public function process_bnpl_return() {
+		$request = Utils::get_request();
+		try {
+			$this->validate_request( $request );
+
+			$gateway_response = $this->gateway->get_transaction_details_by_txn_id( $request->get_param( 'id' ) );
+			$order = $this->get_order( $gateway_response );
+
+			switch( $gateway_response->transactionStatus ) {
+				case TransactionStatus::INITIATED:
+					wp_safe_redirect( $order->get_checkout_order_received_url() );
+					break;
+				case TransactionStatus::PREAUTHORIZED:
+					$note_text = sprintf(
+						'%1$s %2$s. Transaction ID: %3$s.',
+						wc_price( $order->get_total() ),
+						__( 'authorized', 'globalpayments-gateway-provider-for-woocommerce' ),
+						$order->get_transaction_id()
+					);
+					$order->update_status( 'processing', $note_text );
+
+					if ( $this->payment_action == AbstractGateway::TXN_TYPE_SALE ) {
+						AbstractGateway::capture_credit_card_authorization( $order );
+					}
+					$order->payment_complete();
+					wp_safe_redirect( $order->get_checkout_order_received_url() );
+					break;
+				case TransactionStatus::DECLINED:
+				case 'FAILED':
+					$note_text = sprintf(
+						'%1$s %2$s. Transaction ID: %3$s.',
+						wc_price( $order->get_total() ),
+						__( 'declined', 'globalpayments-gateway-provider-for-woocommerce' ),
+						$order->get_transaction_id()
+					);
+					$order->update_status( 'failed', $note_text );
+					wc_add_notice( Utils::map_response_code_to_friendly_message( $gateway_response->transactionStatus ), 'error' );
+					wp_safe_redirect( wc_get_checkout_url() );
+					break;
+				default:
+					throw new \Exception(
+						'Order ID: ' . $gateway_response->orderId . '. Unexpected transaction status on returnUrl: ' . $gateway_response->transactionStatus
+					);
+			}
+		} catch ( \Exception $e ) {
+			$log_text = sprintf(
+				'Error completing order return with ' . $this->payment_method_BNPL_provider . '. %s %s',
+				$e->getMessage(),
+				print_r( $request->get_params(), true )
+			);
```

Comment:
In this case, we should show a message to the user, just like in the DECLINED/FAILED case.